src/jit/compiler.cpp

```diff
@@ -11097,3 +11097,831 @@ bool Compiler::killGCRefs(GenTree* tree)
+             it++)
+        {
+            dumpVariableLiveRange(&it, codeGen);
+        }
+        printf("]\n");
+    }
+    else
+    {
+        printf("None history\n");
+    }
+}
+
+// Returns true if a live range for this variable has been recorded from last call to EndBlock
+bool VariableLiveDescriptor::hasVarHomesToDump() const
+{
+    // "m_VariableLifeBarrier" should has been initialized
+    noway_assert(m_VariableLifeBarrier != nullptr);
+
+    return m_VariableLifeBarrier->hasLiveRangesToDump();
+}
+
+// Reset the barrier so as to dump only next block changes on next block
+void VariableLiveDescriptor::endBlockLiveRanges()
+{
+    noway_assert(m_VariableLifeBarrier != nullptr);
+
+    // make "m_VariableLifeBarrier->m_StartingLiveRange" now points to nullptr for printing purposes
+    m_VariableLifeBarrier->resetDumper(m_VariableLiveRanges);
+}
+
+#endif // DEBUG
+
+//------------------------------------------------------------------------
+//                      VariableLiveKeeper
+//------------------------------------------------------------------------
+
+void Compiler::initializeVariableLiveKeeper()
+{
+    CompAllocator allocator = getAllocator(CMK_VariableLiveRanges);
+
+    int amountTrackedVariables = opts.compDbgInfo ? info.compLocalsCount : 0;
+    int amountTrackedArgs      = opts.compDbgInfo ? info.compArgsCount : 0;
+
+    varLiveKeeper = allocator.allocate<VariableLiveKeeper>(1);
+    memset(varLiveKeeper, 0, sizeof(*varLiveKeeper));
+    new (varLiveKeeper, jitstd::placement_t()) VariableLiveKeeper(amountTrackedVariables, amountTrackedArgs, this);
+}
+
+VariableLiveKeeper* Compiler::getVariableLiveKeeper() const
+{
+    return varLiveKeeper;
+};
+
+//------------------------------------------------------------------------
+// VariableLiveKeeper: Create an instance of the object in charge of managing
+//  VariableLiveRanges and intialize the array "lvaLiveDsc".
+//
+// Arguments:
+//    totalLocalCount   - the amount of args, special args and IL Local
+//      variables in the method.
+//    argsCount         - the amount of args and special args in the method.
+//    compiler          - a compiler instance
+//
+VariableLiveKeeper::VariableLiveKeeper(unsigned int totalLocalCount, unsigned int argsCount, Compiler* comp)
+{
+    // Fill internal properties
+    lvLiveDscCount              = totalLocalCount;
+    lvLiveArgsCount             = argsCount;
+    compiler                    = comp;
+    lastBasicBlockHasBeenEmited = false;
+
+    if (0 < lvLiveDscCount)
+    {
+        // Get the allocator used for all the VariableLiveRange objects
+        CompAllocator allocator = compiler->getAllocator(CMK_VariableLiveRanges);
+
+        // Allocate memory for "lvaLiveDsc" and initialize each "VariableLiveDescriptor"
+        lvaLiveDsc            = allocator.allocate<VariableLiveDescriptor>(lvLiveDscCount);
+        size_t lvaLiveDscSize = sizeof(*lvaLiveDsc);
+        memset(lvaLiveDsc, 0, lvaLiveDscSize);
```

Comment:
=> new VariableLiveDescriptor[lvLiveDscCount];

?

Then do you even need the loop below?

Reply:
Unfortunately that doesn't work because you need to pass the allocator to `VariableLiveDescriptor`. But still, this code is way to convoluted. Should be something like:
```C++
        lvaLiveDsc = allocator.allocate<VariableLiveDescriptor>(lvLiveDscCount);
        for (unsigned varNum = 0; varNum < lvLiveDscCount; varNum++)
            new (lvaLiveDsc + varNum, jitstd::placement_t()) VariableLiveDescriptor(allocator);
```
Hrm, need to look into getting rid of `jitstd::placement_t()`...

Reply:
Sorry, I thought it was a common practice cleaning the bits of memory (memset). I guess the object should be responsible of setting its default state on its own ctor.

Reply:
In general yes, constructors are supposed to initialize the object as needed, without assuming that the memory has been zeroed out. There may be cases in the JIT where memory is explicitly initialized but that's not typical C++ practice.